### ISSUE TITLE
twisted.plugin.getCache fails on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11075]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| amberite reported|
|-|-|
|Trac ID|trac#1445|
|Type|defect|
|Created|2006-01-24 05:39:34Z|

Attachments:

* [patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b85/b8509b02cc4c001875e9fcb20f3c7c0df970b5a0/9c2db015e47d179fd5565f2f2144ace7e021399c.diff) (802 bytes) - added by amberite on 2006-01-24 05:39:22Z - 
* [patch.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b85/b8509b02cc4c001875e9fcb20f3c7c0df970b5a0/15518c5150feb4374a1fcf8970de9b7502f62d58.diff) (1057 bytes) - added by amberite on 2006-01-24 06:25:40Z - 
* [plugin-error-handling.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b85/b8509b02cc4c001875e9fcb20f3c7c0df970b5a0/f971ce6fe742e780047afe39462efd324cbd115e.patch) (1329 bytes) - added by exarkun on 2006-02-07 10:00:54Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1445 1445
type__defect defect
reporter__amberite amberite
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__win32 win32
time__1138081174000000 1138081174000000
changetime__1229063792000000 1229063792000000
version__ 
owner__ 
cc__exarkun cc__amberite cc__teratorn
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
The dropin.cache file never gets updated properly, due to the fact that, on Windows, os.
rename does not work on existing files.

A brief chat with exarkun suggests that this could be handled by wrapping the call to os.
rename in a try/except block, and if the error is errno.EEXIST, calling os.remove on the 
existing file.

Patch attached, please discuss.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
Updated patch to use the "log.msg, then log.err()" idiom, in the hopes that it will cause 
the associated trial test to fail.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Slightly modified patch attached.

Unfortunately, it causes all the plugin tests to fail on my machine.  Not sure
why right now.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @teratorn|
|-|-|

I am not a Windows guy.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn set status to `assigned`|
|-|-|

Thanks guys, I will have a look at this. As we have done before, a simple delete/rename scheme should work fine.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn removed owner<br>@teratorn set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

It looks like this works already.  I found this bug because I was removing a bunch of duplicate code from that module :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [19199]) switch twisted.plugin to use twisted.python.modules re [#11075](https://github.com/twisted/twisted/issues/11075) re [#1951](https://github.com/twisted/twisted/issues/1951)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set status to `closed`|
|-|-|

This was actually fixed in r19305, when [#1951](https://github.com/twisted/twisted/issues/1951) was fixed.  I tested interactively to confirm, but I believe that `twisted.test.test_plugin.PluginTestCase.test_detectFilesChanged` (and perhaps some others), which was fixed in r21907 as part of the de-globalification of the plugins tests, would be failing on a Windows buildbot if this were still broken.
